packages/query/src/block-processor.ts

```diff
@@ -207,6 +150,9 @@ export class BlockProcessor implements BlockProcessorInterface {
       keepAlive: true,
       abortSignal: this.abortController.signal,
     })) {
+      if (compactBlock.height !== ++currentHeight) {
+        throw new Error(`Unexpected block: Recieved ${compactBlock.height} at ${currentHeight}`);
+      }
```

Comment:
What are the cases where this would happen? Why would startHeight ever be out of sync with what the response returns?

Is the bug that sometimes the query does not respect the startHeight we pass?

Reply:
Does throwing an error correct the issue? Aka, after retry, would it be fixed?

Reply:
Having the check is a good defense in depth. You are correct that the height should be set by the request. However, we built a machine that corrupts RPC responses and put it in front of the default testnet RPC.

None of the linked issues have investigated whether there is an actual bug in pd or not. This would be a good next step.

Alternatively, we could avoid corrupting RPC responses just by not choosing to load balance requests over nodes that have inconsistent views of the chain state.

Reply:
i and others have seen the issue in action.

after throwing, the automatic retry will reset the state and make another request. this will resolve the issue if the remote endpoint returns the expected start height.

if the remote endpoint does not return the expected start height, the user's local state will fail to sync.